### DISCUSSION TITLE
Unable to define multiple primary keys

### DISCUSSION THREAD

**Discussion by ghost** [mirumee/ariadne#494] [Q&A]
The Apollo Federation docs (and apollo-server), have a feature to [define multiple primary keys][1]. This allows different services resolve the entity by different fields. Here is the example from the Apollo docs:

```
type Product @key(fields: "upc") @key(fields: "sku") {
  upc: String!
  sku: String!
  price: String
}
```

Running this in ariadne (I am running `0.12.0`), gives me the error: (the code to replicate is below)

```
TypeError: The directive 'key' can only be used once at this location.
```

In my case, I am happy to steer clear of this feature, however, our other microservices (running on Node) take advantage of this and there may come a time we want to take advantage of these types.

_Code to replicate:_

```
# schema file
from ariadne.contrib.federation import make_federated_schema

type_defs = """
type Query
type Mutation

type Product @key(fields: "upc") @key(fields: "sku") {
  upc: String!
  sku: String!
  price: String
}
"""

schema = make_federated_schema(type_defs)

...

# Flask server

@app.route("/graphql", methods=["POST"])
def graphql_server():
    # GraphQL queries are always sent as POST
    data = request.get_json()

    # Note: Passing the request to the context is optional.
    # In Flask, the current request is always accessible as flask.request
    success, result = graphql_sync(
        schema,
        data,
        context_value=request,
        debug=app.debug
    )

    status_code = 200 if success else 400
    return jsonify(result), status_code
```

[1]:https://www.apollographql.com/docs/apollo-server/federation/entities/#defining-multiple-primary-keys

**Comment by delyanr** [comment]
The `graphql-js` specs, which is what `graphql-core` is based off, do not allow repeatable directives on OBJECT. You can see this clearly in the following [tests](https://github.com/APIs-guru/graphql-js/blob/736a8357e65853b797bb769ea322cf1956113d5f/src/validation/__tests__/UniqueDirectivesPerLocationRule-test.js).

I suggest you use the following instead:

```gql
type Product @key(fields: "upc sku") {
  upc: String!
  sku: String!
  price: String
}
```

**Comment by mjfaga** [comment]
Hi there, 

Wanted to check in on this despite it having been closed last year. The GQL spec does support `repeatable` directives, as outlined by [this test](https://github.com/APIs-guru/graphql-js/blob/736a8357e65853b797bb769ea322cf1956113d5f/src/validation/__tests__/UniqueDirectivesPerLocationRule-test.js#L86-L92).

For example, in a NodeJS federated implementation, the following SDL is valid, but this does not work currently in this library:
```gql
type Product @key(fields: "sku") @key(fields: "id") {
  id: ID!
  sku: String!
  price: String
}
```

Any chance you'd be willing to reconsider to support this functionality?

**Reply by sMorm** [reply]
Just want to add to this, `repeatable` is part of the graphql-spec that has been approved. Here is the ([RFC](https://github.com/graphql/graphql-spec/pull/472)) and the [pre-release draft spec](https://spec.graphql.org/draft/#sec-Type-System.Directives) to show that repeatable is part of the specs.

Since Ariadne is built off of graphql-core v3, it seems like [is_repeatable directives](https://graphql-core-3.readthedocs.io/en/latest/modules/type.html#id2) are already supported. With that said, I think that by adding `repeatable` to the Ariadne's [base federation schema](https://github.com/mirumee/ariadne/blob/master/ariadne/contrib/federation/schema.py#L31) should allow for repeatable @key directives. I can create an issue and open a PR some time next week.

**Comment by rafalp** [comment]
I'm open to giving this another look.

Repo showing an example of this being done with `apollo-server`s federation would give us a head start.

**Comment by kuchichan** [comment]
Hello everyone, 
I have prepared really tiny PR into Ariadne codebase, I have also prepared an example, that mimics apollo federation demo and behaviour described by tests in this [PR](https://github.com/apollographql/apollo-server/pull/2876) (Similar issue in `apollo-server`). Details can be found in: 
https://github.com/kuchichan/ariadne-federation-multiple-keys

**Reply by mjfaga** [reply]
Awesome, you beat me to it, thanks for getting [that PR](https://github.com/mirumee/ariadne/pull/552) up!

**Reply by sMorm** [reply]
Hey @kuchichan — the [federation specs](https://www.apollographql.com/docs/federation/federation-spec/#federation-schema-specification) has been updated to apply `repeatable` to the @key directive, hopefully that will help bolster your changes 😄
